### PR TITLE
HPA: Fix int overflow in GetExternalPerPodMetricReplicas

### PR DESCRIPTION
**Pull request by omerap12**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Addresses potential integer overflows in `GetExternalPerPodMetricReplicas` function to prevent unexpected behavior with extreme metric values.

#### Which issue(s) this PR fixes:

As part of #127022, we've identified a need to address potential integer overflows in our scaling calculations. This PR implements the fix for the `GetExternalPerPodMetricReplicas` function.

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

File: pkg/controller/podautoscaler/replica_calculator.go [hunk 1]
```diff
@@ -369,9 +369,21 @@ func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(statusReplicas int32
 	usageRatio := float64(usage) / (float64(targetUsagePerPod) * float64(replicaCount))
 	if math.Abs(1.0-usageRatio) > c.tolerance {
 		// update number of replicas if the change is large enough
-		replicaCount = int32(math.Ceil(float64(usage) / float64(targetUsagePerPod)))
+		replicaCountResult := math.Ceil(float64(usage) / float64(targetUsagePerPod))
```
Comment (googs1025) [inline pkg/controller/podautoscaler/replica_calculator.go:372]: Should we calculate here first and then use the math.Ceil method? 
like this: 
```go
replicaCountResult := float64(usage) / float64(targetUsagePerPod)
if replicaCountResult > float64(math.MaxInt32) {
   replicaCount = math.MaxInt32
} else {
   replicaCount = math.Ceil(int32(replicaCountResult))
}
```

File: pkg/controller/podautoscaler/replica_calculator_test.go [hunk 2]
```diff
@@ -568,6 +568,38 @@ func TestReplicaCalcScaleUpUnreadyNoScale(t *testing.T) {
 	tc.runTest(t)
 }
 
+func TestExternalPerPodMetricReplicaOverflow(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentReplicas:  1,
+		expectedReplicas: math.MaxInt32,
+		metric: &metricInfo{
+			name:              "qps",
+			levels:            []int64{math.MaxInt64},
```
Comment (googs1025) [inline pkg/controller/podautoscaler/replica_calculator_test.go:577]: I don't quite understand the purpose of this test case. Does our test case test the effect of int overflow?

**Comment by googs1025** [comment]
I'm not quite following the backstory behind this fix. Any chance you could give me a scenario to reproduce it? Also, one thing that's got me curious – when it comes to handling integer overflows, why are we only addressing the maximum overflow and not the minimum? I'm wondering what kind of situations could lead to this issue.
If I've overlooked something, please point it out to me, thanks :)

**Comment by omerap12** [comment]
> I'm not quite following the backstory behind this fix. Any chance you could give me a scenario to reproduce it? Also, one thing that's got me curious – when it comes to handling integer overflows, why are we only addressing the maximum overflow and not the minimum? I'm wondering what kind of situations could lead to this issue. If I've overlooked something, please point it out to me, thanks :)

Hey @googs1025, thanks for your review! We might also want to consider addressing the minimum. Perhaps @sheepster1 could provide us with some additional insights.

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 1 above)
Comment (omerap12) [reply pkg/controller/podautoscaler/replica_calculator.go:372]: Yes. it makes sense. Ill adjust thanks :)

**Comment by sheepster1** [comment]
@googs1025 Wrote about how I encountered this issue over at #126892 
I'm not too familiar with the project. Could you maybe elaborate on what other info can be relevant here?

The exact setup I am using is keda with the Prometheus scaler querying a value larger than maxInt/0.1. I think this can be mocked with a query like this: `vector(300000000)`, and a threshold of 0.1 but I didn't get to testing this mock yet.

**Comment by adrianmoisey** [comment]
I'm not sure I understand what this is fixing. It seems like this piece of code doesn't overflow

**Comment by omerap12** [comment]
> I'm not sure I understand what this is fixing. It seems like this piece of code doesn't overflow

I think you might be right. 
/close

**Comment by sheepster1** [comment]
Are we talking here about the bug report itself? Or this specific bugfix pr?
Here is the original issue: https://github.com/kubernetes/kubernetes/issues/126892

**Comment by adrianmoisey** [comment]
Based on the conversation from here https://github.com/kubernetes/kubernetes/issues/126892#issuecomment-2585649993, Omer is going to keep looking at this
/reopen

**Comment by adrianmoisey** [comment]
This seems fine to me. For some reason is feels like it should be easier, but I can't figure out what that is.
I'll give a LGTM:
/lgtm

**Comment by adrianmoisey** [comment]
@jackfrancis @gjtempleton

File: pkg/controller/podautoscaler/replica_calculator.go [hunk 3]
```diff
@@ -362,16 +362,40 @@ func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(statusReplicas int32
...
+		} else {
+			replicaCount = int32(math.Ceil(replicaCountFloat))
+		}
+	}
+	// Handle usage overflow cases
+	if usage == math.MaxInt64 || float64(usage) >= float64(math.MaxInt64) {
+		usage = math.MaxInt64
+		return replicaCount, usage, timestamp, nil
+	}
+	usageFloat := float64(usage) / float64(statusReplicas)
```
Comment (adrianmoisey) [inline pkg/controller/podautoscaler/replica_calculator.go:388]: I was looking over the backlog and I found this: https://github.com/kubernetes/kubernetes/issues/128847

I wonder if it makes sense to fix in your PR, since you're working in this area.

Comment (omerd-cyera) [reply pkg/controller/podautoscaler/replica_calculator.go:388]: @adrianmoisey I think if you are looking into sorting out the backlog, [this](https://github.com/kubernetes/kubernetes/pull/126979) PR is probably also relevant, since the plan from what i can tell is to continue fixed all the other places after this PR is merged. See https://github.com/kubernetes/kubernetes/pull/126979#issuecomment-2320328068

Comment (omerap12) [reply pkg/controller/podautoscaler/replica_calculator.go:388]: @adrianmoisey, sure I can check it out. But I’d like to get the reviews/approvals opinions first.

File: pkg/controller/podautoscaler/replica_calculator.go [hunk 4]
```diff
@@ -362,16 +362,40 @@ func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(statusReplicas int32
 	}
 	usage = 0
 	for _, val := range metrics {
+		// Check for overflow in summation
+		if val > math.MaxInt64-usage {
```
Comment (jackfrancis) [inline pkg/controller/podautoscaler/replica_calculator.go:366]: Is it possible for `val`s in the `metrics` response slice to be negative values? If not, wouldn't checking for negative values be cheaper and more concise here?

(I assume the answer is yes, otherwise the `usage = usage + val` wouldn't work, but just clarifying.)

File: pkg/controller/podautoscaler/replica_calculator.go [hunk 5]
```diff
@@ -362,16 +362,40 @@ func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(statusReplicas int32
...
-		replicaCount = int32(math.Ceil(float64(usage) / float64(targetUsagePerPod)))
+		replicaCountFloat := float64(usage) / float64(targetUsagePerPod)
+		if replicaCountFloat > float64(math.MaxInt32) {
+			replicaCount = math.MaxInt32
+		} else {
+			replicaCount = int32(math.Ceil(replicaCountFloat))
+		}
+	}
+	// Handle usage overflow cases
+	if usage == math.MaxInt64 || float64(usage) >= float64(math.MaxInt64) {
```
Comment (jackfrancis) [inline pkg/controller/podautoscaler/replica_calculator.go:384]: We can shorten this to just `if float64(usage) >= float64(math.MaxInt64)` as it is inclusive of the equality case

Ref:
- https://play.golang.com/p/GyJdyyKFGEW

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 3 above)
Comment (jackfrancis) [inline pkg/controller/podautoscaler/replica_calculator.go:388]: We already know (from above) that the float64 representation of `usage` is not greater than the max value of the float64 representation of the highest int64 value. In what valid scenarios would dividing the float64 representation of `usage` by the float64 representation of `statusReplicas` yield a higher value, and thus require us to do all of this subsequent overflow validation?

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 4 above)
Comment (omerap12) [reply pkg/controller/podautoscaler/replica_calculator.go:366]: Not sure what you mean here. This overflow check `val > math.MaxInt64-usage` handles both positive and negative values.

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 5 above)
Comment (omerap12) [reply pkg/controller/podautoscaler/replica_calculator.go]: Agree. added in [5d0ec02](https://github.com/kubernetes/kubernetes/pull/127050/commits/5d0ec02ab61c32e1b8bf4e52e00b6bdeaddce1ea)

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 3 above)
Comment (omerap12) [reply pkg/controller/podautoscaler/replica_calculator.go]: Agree. added in [5d0ec02](https://github.com/kubernetes/kubernetes/pull/127050/commits/5d0ec02ab61c32e1b8bf4e52e00b6bdeaddce1ea)

File: pkg/controller/podautoscaler/replica_calculator.go (hunk 4 above)
Comment (jackfrancis) [reply pkg/controller/podautoscaler/replica_calculator.go:366]: What I meant was: would this be a cheaper solution?:

```golang
usage = 0
for _, val := range metrics {
    usage = usage + val
    if usage < 0 {
        // the only way we would ever get here is because of an Int64 overflow
        usage = math.MaxInt64
        break
    }
}
```

Comment (omerap12) [reply pkg/controller/podautoscaler/replica_calculator.go]: Oh ok I get it. agree. Added in [73816bf](https://github.com/kubernetes/kubernetes/pull/127050/commits/73816bf35a3706e74abf1c75918d62fc565ead04)

**Comment by jackfrancis** [comment]
I wonder if @dgrisonnet has any thoughts here on any existing idioms to manage type overflows in metrics summarization calculations. For context, @omerap12 is seeing overflows in summarization metrics that result in unexpected calculation outcomes (in his case an HPA calculation yields a sudden, unexpected scale down when this happens on a cluster at scale).

This seems like a sane way to handle this use case (essentially we are enforcing a ceiling beyond which don't scale out), but curious if others have thoughts.

Also @omerap12 it would be helpful (for me at least) to add these overflow test cases to the existing k/k implementation. Showing the current outcomes would be a convincing argument for "we should fix this". Does that make sense? Like a simple PR against k/k master that adds passing test cases with non-sensical outcomes, to make a point.

cc @adrianmoisey

**Comment by dgrisonnet** [comment]
> I wonder if @dgrisonnet has any thoughts here on any existing idioms to manage type overflows in metrics summarization calculations.

This is my first time encountering this type of issue, usually int64/float64 are more than enough to store the data.

I wouldn't expect valid data to exceed that limit, could the initial issue have been caused by erroneous data? If that's the case, my recommendation would be to somehow surface that info to the users while maxing out the replicas count and avoid at all cost the count to decrease because of an overflow.

**Comment by adrianmoisey** [comment]
> Also @omerap12 it would be helpful (for me at least) to add these overflow test cases to the existing k/k implementation. Showing the current outcomes would be a convincing argument for "we should fix this". Does that make sense? Like a simple PR against k/k master that adds passing test cases with non-sensical outcomes, to make a point.

That's a good idea. The original issue (https://github.com/kubernetes/kubernetes/issues/126892) speaks about the situation required to cause this issue. I've semi-tested this locally, and it seems to be a real bug, but I agree that a test demonstrating the users scenario could be helpful

**Comment by sheepster1** [comment]
> This is my first time encountering this type of issue, usually int64/float64 are more than enough to store the data.

@dgrisonnet the issue here is an int32 overflow not int64. So it happens and much more "realistic" numbers.
Also, I think setting it to max int32 is reasonable here

**Comment by omerap12** [comment]
> Also @omerap12 it would be helpful (for me at least) to add these overflow test cases to the existing k/k implementation. Showing the current outcomes would be a convincing argument for "we should fix this". Does that make sense? Like a simple PR against k/k master that adds passing test cases with non-sensical outcomes, to make a point.
> 
Yes, I understand. That seems reasonable to me. I'll raise a PR.

**Review by hashim21223445 — APPROVED** [review]

File: pkg/controller/podautoscaler/replica_calculator_test.go [hunk 6]
```diff
@@ -561,6 +561,38 @@ func TestReplicaCalcScaleUpUnreadyNoScale(t *testing.T) {
...
+			perPodTargetUsage: 1, // Set to a very small value to force replica count overflow
```
[the remark this answers is not in the document]
Comment (jackfrancis) [reply pkg/controller/podautoscaler/replica_calculator_test.go:571]: Chatted w/ @omerap12 offline, I got this exactly wrong. :)

Could we change this to

`// Maximum distibution of load usage across pods to max out replicas to input levels at 1:1 ratio`

[the remark this answers is not in the document]
Comment (jackfrancis) [reply pkg/controller/podautoscaler/replica_calculator_test.go:571]: maybe we drop the "at 1:1 ratio" part as the replica data type is Int32 and levels data type is Int64

[the remark this answers is not in the document]
Comment (omerap12) [reply pkg/controller/podautoscaler/replica_calculator_test.go]: After discussing with Jack, adjusted the comments accordingly.

[the remark this answers is not in the document]
Comment (jackfrancis) [reply pkg/controller/podautoscaler/replica_calculator_test.go]: thanks!

**Comment by jackfrancis** [comment]
looks like @mwielgus will have to approve

we might want to update [pkg/controller/podautoscaler/OWNERS](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/OWNERS)

cc @gjtempleton

**Comment by omerap12** [comment]
T

> looks like @mwielgus will have to approve
> 
> we might want to update [pkg/controller/podautoscaler/OWNERS](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/OWNERS)
> 
> cc @gjtempleton

Raised #131292 to add @gjtempleton to the approvers but I'm not about its status.

**Comment by omerap12** [comment]
@soltysh , PTAL :)

**Comment by BenTheElder** [comment]
Can we please squash down these commits? Prefer not to have commits like "rebase" or "merged" and to only have logically distinct commits or a single commit.

It makes bisecting kubernetes periodic test failures harder and it makes cherry-picking messier.

**Comment by omerap12** [comment]
> Can we please squash down these commits? Prefer not to have commits like "rebase" or "merged" and to only have logically distinct commits or a single commit.
> 
> It makes bisecting kubernetes periodic test failures harder and it makes cherry-picking messier.

Yes, sorry. fixed :)

**Review by soltysh — APPROVED** [review]
/approve

**Comment by soltysh** [comment]
> @soltysh , PTAL :)

Sorry, just found this. Next time if you want me to get to those PRs faster feel free to reach out to me on slack. That's the agreement and directions I've given to Guy as well :wink:

**Comment by soltysh** [comment]
Flake https://github.com/kubernetes/kubernetes/issues/121751
/test pull-kubernetes-unit

**Comment by omerap12** [comment]
> > @soltysh , PTAL :)
> 
> Sorry, just found this. Next time if you want me to get to those PRs faster feel free to reach out to me on slack. That's the agreement and directions I've given to Guy as well 😉

Noted, thanks! :)

**Comment by omerap12** [comment]
Flake https://github.com/kubernetes/kubernetes/issues/131545
/test pull-kubernetes-e2e-gce
